plugins/mongodb/mongodbembed.hpp

```diff
@@ -0,0 +1,696 @@
+                            }
+                            query++;
+                        }
+                    }
+                }
+                break;
+            default:
+                if (!inStr && ch==searchFor)
+                    return query;
+                break;
+            }
+        }
+        return NULL;
+    }
+
+    static unsigned countParameterPlaceholders(const char *query)
+    {
+        unsigned queryCount = 0;
+        while ((query = findUnquotedChar(query, '$')) != NULL)
+            queryCount++;
+        return queryCount;
+    }
+
+    enum PathNodeType {CPNTScalar, CPNTDataset, CPNTSet};
+
+    struct PathTracker
+    {
+        StringBuffer nodeName;
+        PathNodeType nodeType;
+        unsigned int currentChildIndex;
+        unsigned int childCount;
+        unsigned int childrenProcessed;
+
+        // Simple constructor
+        PathTracker()
+        {}
+
+        // Constructor given node name and dataset bool
+        PathTracker(const StringBuffer& _nodeName, PathNodeType _nodeType)
+            : nodeName(_nodeName), nodeType(_nodeType), currentChildIndex(0), childCount(0), childrenProcessed(0)
+        {}
+
+        // Copy constructor
+        PathTracker(const PathTracker& other)
+            : nodeName(other.nodeName), nodeType(other.nodeType), currentChildIndex(other.currentChildIndex), childCount(other.childCount), childrenProcessed(other.childrenProcessed)
+        {}
+    };
+
+    /**
+     * @brief Holds information about where to send the query, how to query, and the results of the query.
+     * 
+     */
+    class MongoDBQuery
+    {
+    public:
+        /**
+         * @brief Stores the databaseName and CollectionName for executing the operations.
+         * 
+         * @param database MongoDB database to connect to.
+         * @param collection MongoDB collection to connect to.
+         */
+        MongoDBQuery(const char *database, const char *collection, std::int32_t _batchSize) 
+        {
+            databaseName = database;
+            collectionName = collection;
+            batchSize = _batchSize;
+        }
+
+        /**
+         * @brief Set the Embed object and remove leading characters.
+         * 
+         * @param chars Size of the script.
+         * @param eScript pointer to beginning of the script.
+         */
+        void setEmbed(size32_t &chars, const char *eScript) 
+        {
+            const char *script = eScript;
+            const char *end;
+
+            // Leading whitespace, new line characters, and tabs are removed.
+            while(*script == ' ' || *script == '\n' || *script == '\t')
+            {
+                script++;
+                chars--;
+            }
+
+            if(embeddedScript.find("(") == std::string::npos) failx("Syntax Error: missing opening parenthesis");
+            end = script;
+            while(*end && (end - script) <= chars && *end != '(') end++; // Get pointer to end of the command
```

Comment:
Confusion over whether the parameter is a null terminated string, or a string with a length.

Reply:
Made the condition clearer as it is a null terminated string.